File: JS/HashMap/ValidSudoko.js
module.exports = {
	//param A : array of strings
	//return an integer
	isValidSudoku : function(A){
        var sudoko = [];

        for(var j = 0; j < 9; j++){
            var rowArr = A[j].split('');
            sudoko.push(rowArr);
            var Set = {};
            for(var i = 0; i < 9; i++){
                var c = rowArr[i];
                //console.log(c);
                if(Set.hasOwnProperty(c))return 0;
                else if(c != '.')Set[c] = true;
            }
        }

        for(var col = 0; col < 9; col++){
            var Set = {};
            for(var row = 0; row < 9; row++){
                var c = sudoko[row][col];
                if(Set.hasOwnProperty(c))return 0;
                else if(c != '.')Set[c] = true;
            }
        }

         //check small sq.
	    for(var row = 0; row < 3; row++){
	        for(var col = 0; col < 3; col++){
	            var Set = {};
	            for(var i  = 0; i < 3; i++){
	                for(var j = 0; j < 3; j++){
	                    var c = sudoko[3*row+i][3*col+j];
                        if(Set.hasOwnProperty(c))return 0;
                        else if(c != '.')Set[c] = true;
	                }
	            }
	        }
	    }

        return 1;
	}
};

File: Java/HashMap/ValidSudoko.java
public class Solution {
	public int isValidSudoku(final List<String> a) {
	    ArrayList<ArrayList<String>> sudoko = new ArrayList<>();
	    HashSet<String> set = new HashSet<>();

	    //check row
	    for(String row : a){
	        ArrayList<String> list = new ArrayList<>(Arrays.asList(row.split("")));
	        sudoko.add(list);
	        set.clear();
	       // System.out.println("\nrow");
	        for(String c : row.split("")){
	          //  System.out.print(c+" ");
	            if(set.contains(c))return 0;
	            else if(!c.equals("."))set.add(c);
	        }
	    }

	    //check columns
	    for(int col = 0; col < 9; col++){
	        set.clear();
	       // System.out.println("col");
	        for(int row = 0; row < 9; row++){
	            String c = sudoko.get(row).get(col);
	         //   System.out.print(c+" ");
	            if(set.contains(c))return 0;
	            else if(!c.equals("."))set.add(c);
	        }
	    }

	    //check small sq.
	    for(int row = 0; row < 3; row++){
	        for(int col = 0; col < 3; col++){
	            set.clear();
	            //System.out.println("\nSmall");
	            for(int i  = 0; i < 3; i++){
	                for(int j = 0; j < 3; j++){
	                    String c = sudoko.get(3*row+i).get(3*col+j);
	                   // System.out.print(c+" ");
        	            if(set.contains(c))return 0;
        	            else if(!c.equals("."))set.add(c);
	                }
	            }
	        }
	    }
	    return 1;
	}
}
